src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
             }
         }
 
-        private async Task PublishDiagnosticsAsync(Document document)
+        protected virtual async Task PublishDiagnosticsAsync(Document document)
         {
-            var diagnostics = await GetDiagnosticsAsync(document, CancellationToken.None).ConfigureAwait(false);
-            var publishDiagnosticsParams = new PublishDiagnosticParams { Diagnostics = diagnostics, Uri = document.GetURI() };
-            await _jsonRpc.NotifyWithParameterObjectAsync(Methods.TextDocumentPublishDiagnosticsName, publishDiagnosticsParams).ConfigureAwait(false);
+            var fileUriToDiagnostics = await GetDiagnosticsAsync(document, CancellationToken.None).ConfigureAwait(false);
+
+            var publishTasks = fileUriToDiagnostics.Keys.Select(async fileUri =>
+            {
+                var publishDiagnosticsParams = new PublishDiagnosticParams { Diagnostics = fileUriToDiagnostics[fileUri], Uri = fileUri };
+                await _jsonRpc.NotifyWithParameterObjectAsync(Methods.TextDocumentPublishDiagnosticsName, publishDiagnosticsParams).ConfigureAwait(false);
+            });
+
+            await Task.WhenAll(publishTasks).ConfigureAwait(false);
```

Comment:
do we have to be parallel?  This scares me *in proc*.  This can saturate cores, and cause lots of garbage to be created (causing gc pauses/hangs).  in-proc that can seriously degrade the editing experience.  We try to avoid this if possible.

Reply:
Nope - was just trying to be too smart for my own good.  Didn't know it would cause issues.  Will change to serialize it.